### ISSUE TITLE
bug: error when reindexing after upgrading to 0.2

### ISSUE THREAD

**Issue by mertalev** [tensorchord/pgvecto.rs#318]
When calling `REINDEX INDEX` on an HNSW index after upgrading and restarting, it throws the error `internal error: entered unreachable code`. Restarting Postgres again doesn't seem to have an effect. Brute-force search still works.

Steps leading up to this:
```sql
START TRANSACTION;

ALTER DATABASE immich SET search_path TO "$user", public, vectors;
CREATE SCHEMA IF NOT EXISTS vectors; -- this step is required for the ALTER EXTENSION line to work
UPDATE pg_catalog.pg_extension SET extversion = '0.1.11' WHERE extname = 'vectors';
UPDATE pg_catalog.pg_extension SET extrelocatable = true WHERE extname = 'vectors';
ALTER EXTENSION vectors SET SCHEMA vectors;
UPDATE pg_catalog.pg_extension SET extrelocatable = false WHERE extname = 'vectors';
ALTER EXTENSION vectors UPDATE TO '0.2.0-alpha.1';
SELECT pgvectors_upgrade();

COMMIT;

-- throws an error "The extension is upgraded so all index files are outdated." - not sure if intended behavior
SELECT idx_status FROM pg_vector_index_stat WHERE indexname = 'clip_index'; 
```
After restarting:
```sql
ALTER DATABASE immich SET search_path TO "$user", public, vectors;

-- throws a different error "The index is not existing in the background worker."
-- not sure if intended behavior, but we handle this error and continue
SELECT idx_status FROM pg_vector_index_stat WHERE indexname = 'clip_index'; 

REINDEX INDEX clip_index; -- throws 'internal error: entered unreachable code'
```
I also see instances of this in the logs:
```
ERROR:  pgvecto.rs: The index is not existing in the background worker.
ADVICE: Drop or rebuild the index.
CONTEXT:  while vacuuming index "face_index" of relation "public.asset_faces"`
```
This log doesn't show up for `clip_index`, only for `face_index`. The app doesn't call `REINDEX INDEX face_index` since doing this for `clip_index` throws an error first, so my hunch was that it was related to this. But these logs continue to appear after manually calling `REINDEX INDEX face_index`.

Calling `SELECT idx_status FROM vectors.pg_vector_index_stat` throws ```ERROR: called `Result::unwrap()` on an `Err` value: NoSuchType("vector_index_stat")```

I can do some more testing to see if I can find the root cause. I wonder if it's because I ran the first part in a transaction?

Edit: Not running in a transaction produces the same result.

Edit 2: Dropping the two indices makes the `SELECT idx_status FROM vectors.pg_vector_index_stat` query return 0 rows successfully. But attempting to create the indices leads to `internal error: entered unreachable code` as well.

Edit 3: I noticed these logs after running `SELECT pgvectors_upgrade()`:
```
WARNING:  pgvecto.rs is upgraded. Restart PostgreSQL to take effects.
immich_server            | [Nest] 8  - 01/30/2024, 5:32:25 AM    WARN [DatabaseService] 
immich_server            |           The pgvecto.rs extension has been updated to 0.2.0.
immich_server            |           Please restart the Postgres instance to complete the update.
immich_postgres          | 2024-01-30 05:32:25.673 UTC [41] ERROR:  pgvecto.rs: The extension is upgraded so all index files are outdated.
immich_postgres          |      ADVICE: Delete all index files. Please read `https://docs.pgvecto.rs/admin/upgrading.html`.
immich_postgres          | 2024-01-30 05:32:25.673 UTC [41] STATEMENT:  COMMIT
immich_postgres          | 2024-01-30 05:32:25.675 UTC [41] ERROR:  pgvecto.rs: The extension is upgraded so all index files are outdated.
immich_postgres          |      ADVICE: Delete all index files. Please read `https://docs.pgvecto.rs/admin/upgrading.html`.
immich_postgres          | 2024-01-30 05:32:25.675 UTC [41] WARNING:  AbortTransaction while in ABORT state
immich_postgres          | 2024-01-30 05:32:25.676 UTC [41] ERROR:  pgvecto.rs: The extension is upgraded so all index files are outdated.
immich_postgres          |      ADVICE: Delete all index files. Please read `https://docs.pgvecto.rs/admin/upgrading.html`.
immich_postgres          | 2024-01-30 05:32:25.676 UTC [41] WARNING:  AbortTransaction while in ABORT state
immich_postgres          | 2024-01-30 05:32:25.677 UTC [41] LOG:  could not send data to client: Broken pipe
immich_postgres          | 2024-01-30 05:32:25.677 UTC [41] ERROR:  pgvecto.rs: The extension is upgraded so all index files are outdated.
immich_postgres          |      ADVICE: Delete all index files. Please read `https://docs.pgvecto.rs/admin/upgrading.html`.
immich_postgres          | 2024-01-30 05:32:25.677 UTC [41] WARNING:  AbortTransaction while in ABORT state
immich_postgres          | 2024-01-30 05:32:25.678 UTC [41] ERROR:  pgvecto.rs: The extension is upgraded so all index files are outdated.
immich_postgres          |      ADVICE: Delete all index files. Please read `https://docs.pgvecto.rs/admin/upgrading.html`.
immich_postgres          | 2024-01-30 05:32:25.678 UTC [41] PANIC:  ERRORDATA_STACK_SIZE exceeded
immich_postgres          | 2024-01-30 05:32:25.679 UTC [1] LOG:  server process (PID 41) was terminated by signal 6: Aborted
immich_postgres          | 2024-01-30 05:32:25.679 UTC [1] DETAIL:  Failed process was running: COMMIT
immich_postgres          | 2024-01-30 05:32:25.679 UTC [1] LOG:  terminating any other active server processes
immich_postgres          | 2024-01-30 05:32:25.681 UTC [1] LOG:  all server processes terminated; reinitializing
immich_postgres          | 2024-01-30 05:32:25.690 UTC [47] LOG:  database system was interrupted; last known up at 2024-01-30 05:32:23 UTC
immich_postgres          | 2024-01-30 05:32:25.786 UTC [48] FATAL:  the database system is in recovery mode
immich_postgres          | 2024-01-30 05:32:28.195 UTC [47] LOG:  database system was not properly shut down; automatic recovery in progress
immich_postgres          | 2024-01-30 05:32:28.204 UTC [47] LOG:  redo starts at 0/7A00118
immich_postgres          | 2024-01-30 05:32:28.209 UTC [47] LOG:  invalid record length at 0/7ADC768: wanted 24, got 0
immich_postgres          | 2024-01-30 05:32:28.209 UTC [47] LOG:  redo done at 0/7ADC708 system usage: CPU: user: 0.00 s, system: 0.00 s, elapsed: 0.00 s
immich_postgres          | 2024-01-30 05:32:28.354 UTC [49] FATAL:  the database system is in recovery mode
immich_postgres          | 2024-01-30 05:32:28.498 UTC [1] LOG:  database system is ready to accept connections
```
This is with running each command as a separate transaction so I *think* the `STATEMENT:  COMMIT` is just referring to `SELECT pgvectors_upgrade()`? I'm not sure.

There are some interesting logs there, especially `PANIC:  ERRORDATA_STACK_SIZE exceeded`, `FATAL:  the database system is in recovery mode` and `invalid record length at 0/7ADC768: wanted 24, got 0`. I think this is probably the root cause.

Altering columns to `real[]`, dropping the indices and extension, re-creating the extension and re-applying the vector type and indices does work.

**Comment by usamoi** [comment]
`ALTER DATABASE` take effects in new sessions so you may need:

```sql
ALTER DATABASE immich SET search_path TO "$user", public, vectors;
SET search_path TO "$user", public, vectors;
```

`The index is not existing in the background worker.` is intended behavior and you may need to reindex them first.

The custom type in PostgreSQL (`vector`) requires no uninitialized bytes during construction. We previously overlooked this and have corrected it in version `0.1.12`, also adding a check for any uninitialized bytes in the data. This check has caused an issue: older data is now considered corrupt by `0.2.0`. I'm not sure if we should provide a patch to fix it automantically.

This does not affect altering the column to `real[]` and reinstalling the extension (setting `search_path` and executing `SELECT pgvectors_upgrade()` are still required). So we suggest this way now. We apologize for any inconvenience caused.

cc @VoVAllen

**Comment by usamoi** [comment]
We provide a patch to fix it. Now the upgrade process should be:

1.

```sql
CREATE SCHEMA IF NOT EXISTS vectors;
ALTER DATABASE immich SET search_path TO "$user", public, vectors;
SET search_path TO "$user", public, vectors;

UPDATE pg_catalog.pg_extension SET extversion = '0.1.11' WHERE extname = 'vectors';
UPDATE pg_catalog.pg_extension SET extrelocatable = true WHERE extname = 'vectors';
ALTER EXTENSION vectors SET SCHEMA vectors;
UPDATE pg_catalog.pg_extension SET extrelocatable = false WHERE extname = 'vectors';
ALTER EXTENSION vectors UPDATE TO '0.2.0-alpha.2';

SELECT pgvectors_upgrade();
```

2. restart PostgreSQL

3. Get all vector indexes.

```sql
SELECT
    C.relname AS tablename,
    I.relname AS indexname
FROM pg_class C JOIN
    pg_index X ON C.oid = X.indrelid JOIN
    pg_class I ON I.oid = X.indexrelid JOIN
    pg_am A ON A.oid = I.relam
WHERE A.amname = 'vectors';
```

```text
 tablename | indexname 
-----------+-----------
t          | t_vector_idx
```

For every row, if the table name is `t`, the column name is `vector` and the indexname is `t_vector_idx`, execute:

```sql
ALTER TABLE t ALTER COLUMN vector SET DATA TYPE real[];
ALTER TABLE t ALTER COLUMN vector SET DATA TYPE vector;
REINDEX t_vector_idx;
```

4. Everything should works now.
